servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java

```diff
@@ -226,8 +226,27 @@ private DnsDiscoveryObserver newDiscoveryObserver(final String address) {
     public Publisher<Collection<ServiceDiscovererEvent<InetAddress>>> dnsQuery(final String address) {
         requireNonNull(address);
         return defer(() -> {
-            ARecordPublisher pub = new ARecordPublisher(address, newDiscoveryObserver(address));
-            return inactiveEventsOnError ? recoverWithInactiveEvents(pub, false) : pub;
+            final DnsDiscoveryObserver discoveryObserver = newDiscoveryObserver(address);
+            ARecordPublisher pub = new ARecordPublisher(address, discoveryObserver);
+            Publisher<? extends Collection<ServiceDiscovererEvent<InetAddress>>> events = inactiveEventsOnError ?
+                    recoverWithInactiveEvents(pub, false) :
+                    pub;
+            return events
+                    .afterCancel(() -> {
+                        if (discoveryObserver != null) {
+                            discoveryObserver.discoveryCanceled();
+                        }
+                    })
+                    .afterOnError(cause -> {
+                        if (discoveryObserver != null) {
```

Comment:
Consider checking this before applying an operator to minimize number of operators on the hot path when there is no observer.